### PR TITLE
DO NOT MERGE - New stereo rendering implementation

### PR DESCRIPTION
**Pull request by jherico**
This PR takes a new approach to stereo rendering.  Rather than calling 'displaySide' twice (and incurring quite a bit of additional CPU side work for each call) we can now switch the rendering context itself into stereo mode.  In stereo mode, a batch that is provided for rendering is actually executed twice, once for each eye, but with slightly different camera parameters.  

Performance testing on master builds without these changes shows that simple side by side rendering of a scene results in a drop to about 40%-50% of the mono framerate.  With these changes, I'm seeing side by side framerates of about 81%-88% of the mono framerate.  In fact, the more CPU intensive a scene is, the better the stereo framerate compared to the mono framerate.  In Cellscience, I'm seeing about 115 FPS in Mono, 102 in SBS Stereo.  Oculus and Vive framerates will still be slightly lower since they render to a large offscreen target, but should still see significant improvement.

Individual batches can have stereo disabled, overriding the context level setting.  This is critical because some of the batches we execute use an entire framebuffer texture as input, so they have to work on the screen as a whole.  

Batches can also set a 'skybox' flag for rendering items at an infinite distance.  If this flag is enabled then the camera only has updated projection information for each eye, and both eyes share the exact same view matrix. 

This PR also removes a number of superfluous syncCache calls.  These calls appeared to be there to re-sync the GPU library state with the actual GL state in cases where the state might have been changed by raw GL.  Since we no longer have any raw GL, these should no longer be necessary.  However, I've left one in at the beginning of the frame (just in case), and also left one in at the beginning of the RenderDefferedTask (because it seems to be necessary, indicating a bug somewhere).  The reduction in the number of syncCache calls appears to improve the framerate by about 5-10%

Known issues:

* Atmosphere rendering is borked in some fashion.  The atmosphere seems to always appear in a fixed part of the screen, rather than a fixed part of the world.
* HMD poses aren't right OR non-HMD poses aren't right.

### CODE REVIEW

File: libraries/display-plugins/src/display-plugins/stereo/StereoDisplayPlugin.cpp [hunk 1]
```diff
@@ -52,6 +52,14 @@ glm::mat4 StereoDisplayPlugin::getModelview(Eye eye, const glm::mat4& baseModelv
 
 void StereoDisplayPlugin::activate() {
     WindowOpenGLDisplayPlugin::activate();
-    CONTAINER->setFullscreen(qApp->primaryScreen());
+//    CONTAINER->setFullscreen(qApp->primaryScreen());
```
Comment (ZappoMan) [inline libraries/display-plugins/src/display-plugins/stereo/StereoDisplayPlugin.cpp]: did you mean to leave this commented out? can you add an explanation why if so.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 97912d2f071914357734253c8879b1fd3a0fdf26 available at: https://gist.github.com/30b6c76d6ef5abb873b8<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a5ed450c1d7f0aeed6507eaa4be67269a9878430 available at: https://gist.github.com/30b6c76d6ef5abb873b8<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 18e12915872b50849f994493800d3a28e58d8417 available at: https://gist.github.com/30b6c76d6ef5abb873b8<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 032aadf1adaec4d670f989b03363a6955935642f available at: https://gist.github.com/30b6c76d6ef5abb873b8<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.dmg

File: libraries/display-plugins/src/display-plugins/stereo/SideBySideStereoDisplayPlugin.cpp [hunk 2]
```diff
@@ -17,7 +17,7 @@
 
 #include <gpu/GLBackend.h>
 
-const QString SideBySideStereoDisplayPlugin::NAME("SBS Stereo Display");
+const QString SideBySideStereoDisplayPlugin::NAME("Debug Stereo Display");
```
Comment (ZappoMan) [inline libraries/display-plugins/src/display-plugins/stereo/SideBySideStereoDisplayPlugin.cpp]: We could call this something other than Debug... after all.. this is actually the mode needed for many 3D TVs...

**Comment by hifi-gustavo** [comment]
Build console output for commit: 84f7f3b34a3455f95ffe0289ddd6366b0c97ed3c available at: https://gist.github.com/30b6c76d6ef5abb873b8<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 892d477da7d13b9179d9e949d56bb2aa0dc0f9d2 available at: https://gist.github.com/30b6c76d6ef5abb873b8<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 35225edae6f511dbd3b1355f1ec4c17e84f22bdf available at: https://gist.github.com/30b6c76d6ef5abb873b8<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2ecf6c6bde8e4587340037c23404bd969d992a3b available at: https://gist.github.com/30b6c76d6ef5abb873b8<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.dmg

**Comment by cozza13** [comment]
I just tried this on the Vive. I had a great 90FPS, however all head movement was opposite. When I physically looked left my head went right etc etc.

**Comment by sethalves** [comment]
I didn't realize it was possible to put a vive HMD on upside-down!

**Comment by hifi-gustavo** [comment]
Build console output for commit: a044eb8c284cfe7fa52adf836737a2974306f9c6 available at: https://gist.github.com/30b6c76d6ef5abb873b8<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.dmg

**Comment by jherico** [comment]
@cozza13 need to make sure that the Oculus and OpenVR plugins are agreeing on the reference frame for all the matrices.

**Comment by hifi-gustavo** [comment]
Build console output for commit: f176178ddef474605c6ec50c4660ea7c4a9eb101 available at: https://gist.github.com/30b6c76d6ef5abb873b8<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5579.dmg

**Comment by ZappoMan** [comment]
We will be working in the STEREO branch to get this read to merge, then open a final merge from there shortly. Closing this for now.

**Pull request closed without merging**